### PR TITLE
In-Memory/JS Storage for msal-browser

### LINKED ISSUE #2166: Error throws when creating msal instance at a third party cookie disabled broswer
>> Please follow the issue template below. Failure to do so will result in a delay in answering your question.

## Library
"@azure/msal-browser": "2.0.0-beta.4",
"msal": "1.3.0",

## Framework
TypeScript/ JavaScript

## Description
This error Happens when:
1. I have a page A running JavaScript call to create a msal instance
2. I have another page B running in a different domain, and trying to put page A inside a iframe
3. Third party cookie has been disabled by browser
I get error throws from the msal package like the Window.sessionStorage is not accessible:
(error from @azure/msal-browser": "2.0.0-beta.4)
image

(error from msal 1.3.0)
image

## Error Message
Failed to read the 'sessionStorage' property from 'Window': Access is denied for this document.
    at BrowserStorage

## Browsers/Environment
- [x] Chrome icongnito when third party cookie disabled
maybe all browser when third party cooked are disabled, it looks like an general problem.

### PR DESCRIPTION
**Pull request by pkanher617**
This PR introduces an in-memory storage option for the `@azure/msal-browser` package. It also adds this as a fallback when local/sessionStorage are not available.

### CODE REVIEW

File: lib/msal-browser/src/cache/InMemoryStorage.ts [hunk 1]
```diff
@@ -0,0 +1,35 @@
+/*
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { IWindowStorage } from "./IWindowStorage";
+
+export class InMemoryStorage implements IWindowStorage {
```
Comment (pkanher617) [inline lib/msal-browser/src/cache/InMemoryStorage.ts:8]: Open to renaming this, as I'm not a huge fan of `InMemoryStorage`.

File: lib/msal-browser/src/cache/InMemoryStorage.ts [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+    private cache: Record<string, string>;
+
+    constructor() {
+        this.cache = {};
```
Comment (jasonnutter) [inline lib/msal-browser/src/cache/InMemoryStorage.ts:13]: Could this be a map?

File: lib/msal-browser/src/cache/BrowserCacheManager.ts [hunk 3]
```diff
@@ -0,0 +1,604 @@
...
+    /**
+     * Gets cache item with given key.
+     * Will retrieve frm cookies if storeAuthStateInCookie is set to true.
+     * @param key
+     */
+    getTemporaryCache(cacheKey: string, generateKey?: boolean): string {
+        const key = generateKey ? this.generateCacheKey(cacheKey) : cacheKey;
+
+        const itemCookie = this.getItemCookie(key);
+        if (this.cacheConfig.storeAuthStateInCookie && itemCookie) {
```
Comment (jasonnutter) [inline lib/msal-browser/src/cache/BrowserCacheManager.ts:293]: Did you pull in the latest from Sameera's branch? This code has been changed.

File: lib/msal-browser/src/utils/BrowserConstants.ts [hunk 4]
```diff
@@ -13,6 +13,8 @@ export const BrowserConstants = {
     CACHE_LOCATION_LOCAL: "localStorage",
     // Session storage constant string
     CACHE_LOCATION_SESSION: "sessionStorage",
+    // In memory storage string
+    CACHE_LOCATION_IN_MEMORY: "inMemoryStorage",
```
Comment (jasonnutter) [inline lib/msal-browser/src/utils/BrowserConstants.ts:17]:
```suggestion
    CACHE_LOCATION_IN_MEMORY: "memoryStorage",
```

Comment (jasonnutter) [reply lib/msal-browser/src/utils/BrowserConstants.ts:17]: And can we create an enum for these strings, and export it?

File: lib/msal-browser/src/cache/BrowserCacheManager.ts [hunk 5]
```diff
@@ -0,0 +1,604 @@
...
+        super();
+
+        this.cacheConfig = cacheConfig;
+        this.clientId = clientId;
+        this.cryptoImpl = cryptoImpl;
+
+        if (cacheConfig.cacheLocation === BrowserConstants.CACHE_LOCATION_IN_MEMORY) {
+            this.browserStorage = new InMemoryStorage();
+        } else {
+            this.browserStorage = new BrowserStorage(cacheConfig.cacheLocation);
```
Comment (jasonnutter) [inline lib/msal-browser/src/cache/BrowserCacheManager.ts:44]: Before this we should check if localStorage/sessionStorage throws.

File: lib/msal-browser/src/cache/BrowserCacheManager.ts (hunk 3 above)
Comment (pkanher617) [reply lib/msal-browser/src/cache/BrowserCacheManager.ts:293]: yeah I think the merge missed it. I will review this file and make sure to add all the changes.

File: lib/msal-browser/src/cache/BrowserCacheManager.ts (hunk 5 above)
Comment (pkanher617) [reply lib/msal-browser/src/cache/BrowserCacheManager.ts:44]: Adding now, that's why this was a draft.

File: lib/msal-browser/src/cache/BrowserCacheManager.ts [hunk 6]
```diff
@@ -0,0 +1,604 @@
...
+    setAccessTokenCredential(accessToken: AccessTokenEntity): void {
+        const accessTokenKey = accessToken.generateCredentialKey();
+        this.setItem(accessTokenKey, JSON.stringify(accessToken));
+    }
+
+    /**
+     * generates refreshToken entity from a string
+     * @param refreshTokenKey
+     */
+    getRefreshTokenCredential(key: string): RefreshTokenEntity {
```
Comment (sameerag) [inline lib/msal-browser/src/cache/BrowserCacheManager.ts:189]: rename the param to `refreshTokenKey`

File: lib/msal-browser/src/cache/BrowserCacheManager.ts [hunk 7]
```diff
@@ -0,0 +1,604 @@
...
+        return accessToken;
+    }
+
+    /**
+     * set accessToken credential to the platform cache
+     * @param accessToken
+     */
+    setAccessTokenCredential(accessToken: AccessTokenEntity): void {
+        const accessTokenKey = accessToken.generateCredentialKey();
+        this.setItem(accessTokenKey, JSON.stringify(accessToken));
```
Comment (sameerag) [inline lib/msal-browser/src/cache/BrowserCacheManager.ts:182]: Can we please add validation for all cache entities?

File: lib/msal-browser/src/cache/BrowserCacheManager.ts [hunk 8]
```diff
@@ -0,0 +1,604 @@
...
+    /**
+     * generates idToken entity from a string
+     * @param idTokenKey
+     */
+    getIdTokenCredential(idTokenKey: string): IdTokenEntity {
+        const value = this.getItem(idTokenKey);
+        if (StringUtils.isEmpty(value)) {
+            return null;
+        }
+        const idToken: IdTokenEntity = CacheManager.toObject(new IdTokenEntity(), JSON.parse(value));
```
Comment (sameerag) [inline lib/msal-browser/src/cache/BrowserCacheManager.ts:150]: add validation

File: lib/msal-browser/src/cache/BrowserCacheManager.ts [hunk 9]
```diff
@@ -0,0 +1,604 @@
...
+        return refreshToken;
+    }
+
+    /**
+     * set refreshToken credential to the platform cache
+     * @param refreshToken
+     */
+    setRefreshTokenCredential(refreshToken: RefreshTokenEntity): void {
+        const refreshTokenKey = refreshToken.generateCredentialKey();
+        this.setItem(refreshTokenKey, JSON.stringify(refreshToken));
```
Comment (sameerag) [inline lib/msal-browser/src/cache/BrowserCacheManager.ts:204]: add validation: `RefreshTokenEntity.isRefreshToken()`

File: lib/msal-browser/src/cache/InMemoryStorage.ts (hunk 2 above)
Comment (sameerag) [reply lib/msal-browser/src/cache/InMemoryStorage.ts:13]: Please make this a map. There is a PR I plan to work on to make all `Record<key, value>` references to `Map` references in our code, in increments. So any new data we are adding which resembles a map, let's make it a map.

File: lib/msal-browser/src/error/BrowserConfigurationAuthError.ts [hunk 10]
```diff
@@ -31,6 +31,10 @@ export const BrowserConfigurationAuthErrorMessage = {
         desc: "The object passed for the callback was invalid. " +
           "More information is available here: https://github.com/AzureAD/microsoft-authentication-library-for-js/wiki/MSAL-basics."
     },
+    inMemRedirectUnavailable: {
+        code: "in_mem_redirect_unavailable",
```
Comment (sameerag) [inline lib/msal-browser/src/error/BrowserConfigurationAuthError.ts:35]: If storeAuthStateInCookie = true and In-Memory is selected, ITP cases still work with v2 right? It is obvious I think but want to check all combos. Also a good idea to make a table of config params and supported functionality as a part of this PR.

File: lib/msal-browser/src/utils/BrowserConstants.ts (hunk 4 above)
Comment (sameerag) [reply lib/msal-browser/src/utils/BrowserConstants.ts:17]: I can have Node expose this too if this is an exported enum and moved to Common!

Comment (jasonnutter) [reply lib/msal-browser/src/utils/BrowserConstants.ts:17]: I think these are just browser considerations that are not relevant for Node.

File: lib/msal-browser/src/error/BrowserConfigurationAuthError.ts (hunk 10 above)
Comment (jasonnutter) [reply lib/msal-browser/src/error/BrowserConfigurationAuthError.ts:35]: Yes, it should, as the cookies created by MSAL are on the app's domain and thus first-party.

File: lib/msal-browser/src/cache/InMemoryStorage.ts (hunk 2 above)
Comment (pkanher617) [reply lib/msal-browser/src/cache/InMemoryStorage.ts]: done.

File: lib/msal-browser/src/cache/BrowserCacheManager.ts (hunk 3 above)
Comment (pkanher617) [reply lib/msal-browser/src/cache/BrowserCacheManager.ts]: done.

File: lib/msal-browser/src/utils/BrowserConstants.ts (hunk 4 above)
Comment (pkanher617) [reply lib/msal-browser/src/utils/BrowserConstants.ts]: done.

File: lib/msal-browser/src/cache/BrowserCacheManager.ts (hunk 5 above)
Comment (pkanher617) [reply lib/msal-browser/src/cache/BrowserCacheManager.ts]: done.

File: lib/msal-browser/src/cache/BrowserCacheManager.ts (hunk 6 above)
Comment (pkanher617) [reply lib/msal-browser/src/cache/BrowserCacheManager.ts]: done.

File: lib/msal-browser/src/cache/BrowserCacheManager.ts (hunk 7 above)
Comment (pkanher617) [reply lib/msal-browser/src/cache/BrowserCacheManager.ts:196]: going to add in a follow up PR, don't want to make this PR larger than it has to be.

File: lib/msal-browser/src/error/BrowserConfigurationAuthError.ts (hunk 10 above)
Comment (pkanher617) [reply lib/msal-browser/src/error/BrowserConfigurationAuthError.ts:35]: will add docs for this in a follow up PR.

File: lib/msal-browser/src/cache/InMemoryStorage.ts (hunk 1 above)
Comment (pkanher617) [reply lib/msal-browser/src/cache/InMemoryStorage.ts]: Changed to MemoryStorage.

File: lib/msal-browser/src/app/ClientApplication.ts [hunk 11]
```diff
@@ -732,12 +733,18 @@ export abstract class ClientApplication {
...
         BrowserUtils.blockNonBrowserEnvironment(this.isBrowserEnvironment);
 
         // block the reload if it occurred inside a hidden iframe
         BrowserUtils.blockReloadInHiddenIframes();
+
+        if (interactionType === InteractionType.Redirect) {
+            if (this.config.cache.cacheLocation === BrowserCacheLocation.MemoryStorage && !this.config.cache.storeAuthStateInCookie) {
+                throw BrowserConfigurationAuthError.createInMemoryRedirectUnavailableError();
+            }
+        }
```
Comment (jasonnutter) [inline lib/msal-browser/src/app/ClientApplication.ts:747]:
```suggestion
        if (interactionType === InteractionType.Redirect && this.config.cache.cacheLocation === BrowserCacheLocation.MemoryStorage && !this.config.cache.storeAuthStateInCookie) {
            throw BrowserConfigurationAuthError.createInMemoryRedirectUnavailableError();
        }
```

File: lib/msal-browser/src/cache/BrowserCacheManager.ts [hunk 12]
```diff
@@ -0,0 +1,617 @@
...
+
+        // Migrate any cache entries from older versions of MSAL.
+        this.migrateCacheEntries();
+    }
+
+    /**
+     * Returns a window storage class implementing the IWindowStorage interface that corresponds to the configured cacheLocation.
+     * @param cacheLocation 
+     */
+    private setupBrowserStorage(cacheLocation: string): IWindowStorage {
```
Comment (jasonnutter) [inline lib/msal-browser/src/cache/BrowserCacheManager.ts:50]:
```suggestion
    private setupBrowserStorage(cacheLocation: BrowserCacheLocation.MemoryStorage): IWindowStorage {
```

File: lib/msal-browser/src/cache/BrowserCacheManager.ts [hunk 13]
```diff
@@ -0,0 +1,617 @@
...
+        if (cacheLocation !== BrowserCacheLocation.MemoryStorage) {
+            try {
+                return new BrowserStorage(cacheLocation);
+            } catch (e) {
+                this.cacheConfig.cacheLocation = BrowserCacheLocation.MemoryStorage;
+                // TODO: Log error here
+            }
+        }
+
+        return new MemoryStorage();
```
Comment (jasonnutter) [inline lib/msal-browser/src/cache/BrowserCacheManager.ts:60]: Potential alternative:

```suggestion
        switch (cacheLocation) {
            case BrowserCacheLocation.LocalStorage:
            case BrowserCacheLocation.SessionStorage:
                try {
                    return new BrowserStorage(cacheLocation);
                } catch (e) {
                    this.cacheConfig.cacheLocation = BrowserCacheLocation.MemoryStorage;
                    return new MemoryStorage();
                }
            case BrowserCacheLocation.MemoryStorage:
                return new BrowserStorage(cacheLocation);
        }
```

File: lib/msal-browser/src/cache/MemoryStorage.ts [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+    setItem(key: string, value: string): void {
+        this.cache.set(key, value);
+    }
+
+    removeItem(key: string): void {
+        this.cache.delete(key);
+    }
+
+    getKeys(): string[] {
+        return [...this.cache.keys()];
```
Comment (jasonnutter) [inline lib/msal-browser/src/cache/MemoryStorage.ts:29]: Not available in IE, unfortunately. Probably good to be safe and manually iteratate over the map and build the array manually.

**Review by jasonnutter — CHANGES_REQUESTED** [review]
A few comments

File: lib/msal-browser/src/cache/BrowserCacheManager.ts (hunk 13 above)
Comment (pkanher617) [reply lib/msal-browser/src/cache/BrowserCacheManager.ts:60]: line 61 should be `return new MemoryStorage();` but I see your point. I was trying to do this with a single `if` statement, which is why it is the way it is. I can change to using switch if it's more readable.

File: lib/msal-browser/src/utils/BrowserConstants.ts (hunk 4 above)
Comment (sameerag) [reply lib/msal-browser/src/utils/BrowserConstants.ts]: Node will have this relevant when we move out of `serializedCache` but it isn't needed now.

File: lib/msal-browser/src/app/ClientApplication.ts (hunk 11 above)
Comment (pkanher617) [reply lib/msal-browser/src/app/ClientApplication.ts]: done.

File: lib/msal-browser/src/cache/BrowserCacheManager.ts (hunk 12 above)
Comment (pkanher617) [reply lib/msal-browser/src/cache/BrowserCacheManager.ts]: done.

File: lib/msal-browser/src/cache/MemoryStorage.ts (hunk 14 above)
Comment (pkanher617) [reply lib/msal-browser/src/cache/MemoryStorage.ts]: done.

File: lib/msal-browser/src/cache/BrowserCacheManager.ts (hunk 13 above)
Comment (pkanher617) [reply lib/msal-browser/src/cache/BrowserCacheManager.ts]: changed to switch statement

File: lib/msal-browser/src/cache/MemoryStorage.ts [hunk 15]
```diff
@@ -0,0 +1,39 @@
+/*
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { IWindowStorage } from "./IWindowStorage";
+
+export class MemoryStorage implements IWindowStorage {
```
Comment (jasonnutter) [inline lib/msal-browser/src/cache/MemoryStorage.ts:8]: Are there tests for this class?

File: lib/msal-browser/src/utils/BrowserConstants.ts [hunk 16]
```diff
@@ -29,6 +25,12 @@ export const BrowserConstants = {
     MSAL_SKU: "msal.js.browser",
 };
 
+export enum BrowserCacheLocation {
```
Comment (jasonnutter) [inline lib/msal-browser/src/utils/BrowserConstants.ts:28]: Can we export this from the main package too?

File: samples/msal-browser-samples/VanillaJSTestApp2.0/app/inMemoryStorage/authConfig.js [hunk 17]
```diff
@@ -0,0 +1,55 @@
+// Config object to be passed to Msal on creation
+const msalConfig = {
+    auth: {
+        clientId: "3fba556e-5d4a-48e3-8e1a-fd57c12cb82e",
+        authority: "https://login.windows-ppe.net/common/"
+    },
+    cache: {
+        cacheLocation: "inMemoryStorage", // This configures where your cache will be stored
```
Comment (jasonnutter) [inline samples/msal-browser-samples/VanillaJSTestApp2.0/app/inMemoryStorage/authConfig.js:8]:
```suggestion
        cacheLocation: "memoryStorage", // This configures where your cache will be stored
```

Comment (jasonnutter) [reply samples/msal-browser-samples/VanillaJSTestApp2.0/app/inMemoryStorage/authConfig.js:8]: Good example of where the enum would be helpful

File: lib/msal-browser/src/utils/BrowserConstants.ts (hunk 16 above)
Comment (pkanher617) [reply lib/msal-browser/src/utils/BrowserConstants.ts:28]: Will changing this config value to an enum affect current implementations?

Comment (pkanher617) [reply lib/msal-browser/src/utils/BrowserConstants.ts:28]: today users pass the string "localStorage" or "sessionStorage"

Comment (jasonnutter) [reply lib/msal-browser/src/utils/BrowserConstants.ts:28]: I think typescript will recognize that the enum values match those values.

File: lib/msal-browser/src/cache/MemoryStorage.ts (hunk 15 above)
Comment (pkanher617) [reply lib/msal-browser/src/cache/MemoryStorage.ts:8]: tests are in BrowserCacheManager

File: lib/msal-browser/src/utils/BrowserConstants.ts (hunk 16 above)
Comment (pkanher617) [reply lib/msal-browser/src/utils/BrowserConstants.ts:28]: done.

File: samples/msal-browser-samples/VanillaJSTestApp2.0/app/inMemoryStorage/authConfig.js (hunk 17 above)
Comment (pkanher617) [reply samples/msal-browser-samples/VanillaJSTestApp2.0/app/inMemoryStorage/authConfig.js]: done.

File: samples/msal-browser-samples/VanillaJSTestApp2.0/app/inMemoryStorage/authConfig.js [hunk 18]
```diff
@@ -0,0 +1,55 @@
+// Config object to be passed to Msal on creation
+const msalConfig = {
```
Comment (tnorling) [inline samples/msal-browser-samples/VanillaJSTestApp2.0/app/inMemoryStorage/authConfig.js:2]: If the purpose of adding this sample is for E2E testing can we just add the tests to the customizable-e2e-test?
Would like to try to reduce the number of samples we have

File: lib/msal-browser/src/cache/BrowserCacheManager.ts (hunk 7 above)
Comment (sameerag) [reply lib/msal-browser/src/cache/BrowserCacheManager.ts:200]: Alright!

**Review by sameerag — APPROVED** [review]

**Review by jasonnutter — APPROVED** [review]

**Comment by jasonnutter** [comment]
Looks good, thanks for addressing feedback!

**Review by tnorling — APPROVED** [review]

File: samples/msal-browser-samples/VanillaJSTestApp2.0/app/customizable-e2e-test/test/browserMemStorage.spec.ts [hunk 19]
```diff
@@ -0,0 +1,192 @@
...
+import { Screenshot, createFolder, setupCredentials } from "../../../../../e2eTestUtils/TestUtils";
+import { BrowserCacheUtils } from "../../../../../e2eTestUtils/BrowserCacheTestUtils";
+import { LabApiQueryParams } from "../../../../../e2eTestUtils/LabApiQueryParams";
+import { AzureEnvironments, AppTypes } from "../../../../../e2eTestUtils/Constants";
+import { LabClient } from "../../../../../e2eTestUtils/LabClient";
+import { msalConfig as memStorageConfig, request as memStorageTokenRequest } from "../authConfigs/memStorageAuthConfig.json";
+import { clickLoginPopup, clickLoginRedirect, enterCredentials, waitForReturnToApp } from "./testUtils";
+import fs from "fs";
+
+const SCREENSHOT_BASE_FOLDER_NAME = `${__dirname}/screenshots`;
```
Comment (tnorling) [inline samples/msal-browser-samples/VanillaJSTestApp2.0/app/customizable-e2e-test/test/browserMemStorage.spec.ts:13]:
```suggestion
const SCREENSHOT_BASE_FOLDER_NAME = `${__dirname}/screenshots/memStorageTests`;
```

**Pull request merged**